pytorch_lightning/core/optimizer.py

```diff
@@ -0,0 +1,253 @@
+                # as they will be aggregated together on epoch_end
+
+                def gen_closure():
+                    ... forward and compute loss for generator
+                    loss_gen = ...
+                    self.log("loss_gen", loss_gen, on_step=True, on_epoch=True)
+                    self.manual_backward(loss_gen, opt_gen)
+
+                def dis_closure():
+                    ... forward and compute loss for discriminator
+                    loss_dis = ...
+                    self.log("loss_dis", loss_dis, on_step=True, on_epoch=True)
+                    self.manual_backward(loss_dis, opt_dis)
+
+                # this will accumulate gradients for 2 batches and then call opt_gen.step()
+                opt_gen.step(closure=gen_closure, make_optimizer_step=batch_idx % 2 == 0)
+
+                # update discriminator every 4 batches
+                # therefore, no gradient accumulation for discriminator
+                if batch_idx % 4 == 0 :
+                    # Note: Set make_optimizer_step to True or it will use by default
+                    # Trainer(accumulate_grad_batches=x)
+                    opt_dis.step(closure=optimizer_closure, make_optimizer_step=True)
+        """
+
+        profiler_name = "optimizer_step_and_closure"
+        if closure is None:
+            closure = do_nothing_closure
+            profile_name = "optimizer_step"
+        else:
+            if not isinstance(closure, types.FunctionType):
+                raise MisconfigurationException("When closure is provided, it should be a function")
+
+        if make_optimizer_step is None:
+            make_optimizer_step = not self._should_accumulate
+
+        trainer = self._trainer
+        optimizer = self._optimizer
+
+        if make_optimizer_step:
+            if trainer.on_tpu:
+                with trainer.profiler.profile(profiler_name):
+                    xm.optimizer_step(optimizer, optimizer_args={'closure': closure, **kwargs})
+
+            elif trainer.amp_backend is not None:
+                trainer.precision_connector.backend.step(trainer, optimizer, closure)
+
+            else:
+                with trainer.profiler.profile(profiler_name):
+                    optimizer.step(closure=closure, *args, **kwargs)
+
+            # perform zero grad
+            optimizer.zero_grad()
+        else:
+            # make sure to call optimizer_closure when accumulating
+            with trainer.train_loop.block_ddp_sync_behaviour():
+                with trainer.profiler.profile("closure"):
+                    closure()
+
+    def __repr__(self):
+        groups = "["
+        for i, group in enumerate(self.param_groups):
+            groups += '('
+            params = ''
+            for key in sorted(group.keys()):
+                if key != 'params':
+                    params += f'{key}={round(group[key], 12)}, '
+            groups += params[:-2]
+            groups += '),'
+        groups = groups[:-1] + ']'
```

Comment:
Why not one-line? :smile: 

```suggestion
        groups = [{k: round(v, 12) for k, v in sorted(group.items()) if k != "params"} for group in param_groups]
```

Reply:
```
LightningOptimizer(optim=SGD, groups=[{'dampening': 0, 'initial_lr': 0.1, 'lr': 0.01, 'momentum': 0, 'nesterov': 0, 'weight_decay': 0}])
```

Reply:
That is missing the repr for the optimizer name. You can change it like this

```python
>>> optim = "SGD"
>>> print(f"{optim}")
SGD
>>> print(f"{optim!r}")
'SGD'
```

In your case:

```python
(optim={self._optimizer.__class__.__name__!r}
```

So in the end the repr function would be (black formatted):


```python
def __repr__(self):
    groups = [
        {
            k: round(v, 12) if isinstance(v, float) else v
            for k, v in sorted(group.items())
            if k != params
        }
        for group in param_groups
    ]
    return f"{self.__class__.__name__}(optim={self._optimizer.__class__.__name__!r}, groups={groups})"
```

Reply:
Thanks !